filemanager.lua

```diff
@@ -175,40 +174,55 @@ function preQuitAll(view) treeView.Buf.IsModified = false end
 
 -- scanDir will scan contents of the directory passed.
 function scanDir(directory)
-    debug("***** scanDir(directory) ---> ",directory)
-    local i, list, proc = 3, {}, nil
-    list[1] = (isWin and driveLetter or "") .. cwd  -- current directory working.
-    list[2] = ".."  -- used for going up a level in directory.
-    if isWin then  -- if windows
-        proc = io.popen('dir /a /b "'..directory..'"')
-    else           -- linux or unix system
-        proc = io.popen('ls -Ap "'..directory..'"')
-    end
-    -- load filenames to a list
-    for filename in proc:lines() do
-        list[i] = filename
-        i = i + 1
+  messenger:AddLog("***** scanDir(directory) ---> ", directory)
+
+  local list = {[1] = cwd, [2] = ".."}
+
+  local go_ioutil = import("ioutil")
+  -- Gets a list of all the files in the current dir
+  local readout = go_ioutil.ReadDir(directory)
+
+  if readout == nil then
+    messenger:Error("Error reading directory: ", directory)
+  else
+    local readout_name = ""
+    -- Loop through all the files/directories in current dir
+    for i = 1, #readout do
+      -- Save the current dir/file name
+      readout_name = readout[i]:Name()
+      -- Check if the current file is a dir
+      if isDir(readout_name) then
+        -- Add on a slash to signify the listing is a directory
+        -- Shouldn't cause issues on Windows, as it lets you use either slash type
+        readout_name = readout_name .. "/"
+      end
+      -- Actually add the file/dir to the list to be displayed
+      list[i + 2] = readout_name
     end
-    proc:close()
-    return list
+  end
+
+  return list
 end
 
 -- isDir checks if the path passed is a directory.
 -- return true if it is a directory else false if it is not a directory.
 function isDir(path)
-    debug("***** isDir(path) ---> ",path)
-    local dir, proc = false, nil
-    if isWin then
-        proc = io.popen('IF EXIST ' .. driveLetter .. JoinPaths(cwd, path) .. '/* (ECHO d) ELSE (ECHO -)')
-    else
-        proc = io.popen('ls -adl "' .. JoinPaths(cwd, path) .. '"')
-    end
-    if proc:read(1) == "d" then
-        dir = true
-    end
-    proc:close()
-    debug("***** isDir return ",dir)
-    return dir
+  debug("***** isDir(path) ---> ", path)
+
+  local go_os = import("os")
+
+  local check_path = JoinPaths(cwd, path)
+
+  -- Returns a FileInfo on the current file/path
+  local file_info = go_os.Stat(check_path)
+
+  if file_info ~= nil then
+    -- Returns the true/false of if the file is a directory
+    return file_info:IsDir()
+  else
+    messenger:AddLog("isDir() failed, returning nil")
+    return nil
```

Comment:
Not sure we should have a nil return without checking in the code where it is used. As this could cause problems if it fails. (nil checking)

Reply:
Everywhere in the code that `isDir` is used is only in `if`'s, so it will simply fail the `if isDir(file)` or whatever. 

Nothing should "explode", as it's basically like saying `false` but with some more verbosity if it ever needs to be checked against more thoroughly.

Reply:
Of course if you're still worried, all `if isDir(file)` lines could be changed to `if isDir(file) and not isDir(file) == nil`, but that seems a bit excessive.

Reply:
@NicolaiSoeborg This is ready to be merged if you are happy with it.
Could not find any fails when testing and it works on Android/Termux so #2 can be closed when merged.

Reply:
I'm very happy! Will merge now.